### PR TITLE
Add loading cmake-kits from a config file

### PR DESCRIPTION
**Pull request by mark2185**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -56,6 +56,7 @@ The current version of the plugin supports next commands:
  - **`:CMakeSelectBuildType`** changes the cmake build type with argument passed and call **`:CMake`**.
  - **`:CMakeInfo`** creates a window with CMake information.
  - **`:CMakeClean`** cleans the project (it is equal of the execution `make clean`).
+ - **`:CMakeKitsReload`** reloads the CMake Kits
```
Comment (ilyachur) [inline README.md:59]: Why do we need to have this additional command? From my point of view the plugin API already overcomplicated ( I have a plan to deprecate at least some options and don't want to introduce new commands if it is not needed).

From my point of view we need to reload cmake kits automatically before `:CMake` command if it is needed.

This new command covers only one specific case when user works with cmake kits. Moreover user can forget to call this command.

Comment (mark2185) [reply README.md:59]: Because then I'd need to implement caching, this was the easier choice.

I don't want to read and parse the JSON file every time the user runs `CMake` if it never changes so I just added the command to _explicitly_ reload the configs.

I will implement the cache, then, not a problem.

File: README.md [hunk 2]
```diff
@@ -116,12 +116,20 @@ Plugin supports special global variables which are allow to change behaviour of
...
  - **`g:cmake_variants`** enables predefined cmake build variants in the form of a dictionary, e.g. `{ 'Debug' : { 'cmake_build_type' : 'Debug', 'cmake_usr_args' : { 'CONAN_PATH' : '~/.conan' } }`
  - **`g:cmake_kits`** enables predefined cmake kits in the form of a dictionary of dictionaries that specify a toolchain file, environment variables, cmake variables among other things
+ - **`g:cmake_kits_global_path`** specifies a path to the JSON containing cmake kits. Default is empty.
  - **`g:cmake_selected_kit`** currently selected cmake kit. Default is empty.
  - **`g:cmake_toolchain_file`** currently selected toolchain file. Default is empty.
  - **`g:cmake_build_path_pattern`** pattern for build dir, two strings that will be evaluated in a `printf`. e.g.:
      `let g:cmake_build_path_pattern = [ "%s/workspace/build/%s/%s/%s", "$HOME, fnamemodify( getcwd(), ':t' ), g:cmake_selected_kit, g:cmake_build_type" ]`, **Note:** it takes precedence over `g:cmake_build_dir_prefix`
 
-Example of supported functions in `g:cmake_kits`:
```
Comment (ilyachur) [inline README.md:124]: Could you rebase your PR to the latest master

File: README.md [hunk 3]
```diff
@@ -228,7 +246,7 @@ endfunction
 function! s:FZFSelectKit() abort
```
Comment (ilyachur) [inline README.md:246]: I propose to remove this example because you already have implemented: `fzf#cmake4vim#SelectKit()`

Moreover could you double check that we don't have redundant examples?

File: autoload/cmake4vim.vim [hunk 4]
```diff
@@ -62,6 +62,9 @@ function! cmake4vim#GenerateCMake(...) abort
     " Prepare requests to CMake system
     call utils#cmake#common#makeRequests(l:build_dir)
 
+    " reload CMakeKits if needed
+    " call utils#cmake#reloadCMakeKits()
```
Comment (ilyachur) [inline autoload/cmake4vim.vim:66]: I propose to remove unused code.

File: autoload/utils/cmake.vim [hunk 5]
```diff
@@ -377,3 +385,7 @@ function! utils#cmake#getBinaryPath() abort
     endif
     return l:exec_path
 endfunction
+
+function! utils#cmake#reloadCMakeKits() abort
```
Comment (ilyachur) [inline autoload/utils/cmake.vim:389]: Looks like we don't use this function.

File: autoload/utils/cmake.vim [hunk 6]
```diff
@@ -88,16 +89,23 @@ function! s:populateDefaultCMakeVariants() abort
...
-    let l:cmake_kit  = get( g:cmake_kits, a:name, {} )
+    let l:cmake_kit = get( s:loaded_cmake_kits, a:name, {} )
     for [key, val] in items( get( l:cmake_kit, 'environment_variables', {} ) )
         execute printf('let $%s="%s"', key, val)
     endfor
 endfunction
 
 function! utils#cmake#unsetEnv(name) abort
-    let l:cmake_kit  = get( g:cmake_kits, a:name, {} )
+    let l:cmake_kit = get( s:loaded_cmake_kits, a:name, {} )
```
Comment (ilyachur) [inline autoload/utils/cmake.vim:108]: Why don't you use `g:cmake_kits`?
What happens if I have  changed this variable in the runtime?

File: autoload/cmake4vim.vim (hunk 4 above)
Comment (mark2185) [reply autoload/cmake4vim.vim:66]: This was for testing because this is the only change to `GenerateCMake` in this PR so I thought commenting this out might make CI work. But it did not help.

File: autoload/utils/cmake.vim (hunk 6 above)
Comment (mark2185) [reply autoload/utils/cmake.vim:108]: Because there are multiple ways to supply `cmake_kits` now.

If you have a `.cmake-kits.json`, that file will be read. No matter what you do to `g:cmake_kits`, as long as you have `.cmake-kits.json` or `g:cmake_kits_global_path`, `g:cmake_kits` won't be used. I added that to `README`.

If you have no files and only use `g:cmake_kits`, then it will reload when you run `CMake` or call `SelectKit` or manually run `:call utils#cmake#reloadCMakeKits()`.

File: README.md (hunk 3 above)
Comment (mark2185) [reply README.md:246]: This is only for the case you want to add more entries into `g:cmake_kits` which are not supported by the plugin.

E.g. if you're using `YouCompleteMe` and developing with `emscripten`, you could pass `em++` to `clangd` via `g:ycm_clangd_args`. But I didn't want to add this functionality to the plugin, because you don't have to use both `cmake4vim` *and* `YouCompleteMe`, but in case you do (like many people do, myself incuded), I wrote these instructions to help people.

And if someone wants to do something like that, you can't override the plugin directly (in `~/.vim/plugged/cmake4vim`, or wherever your plugin is), because then you can't update your plugin, or the update will revert your changes (depends on the plugin manager), so it has to go to the `~/.vim/after` folder, so it gets applied _after_ the plugin initializes.

**Comment by mark2185** [comment]
Okay, this is still buggy, it seems. By putting the `cmake_kits.vader` after `run_target.vader`, everything passes, but I'm not sure why. Something is not reset, I'll have to investigate.

In short - **don't merge yet.**

File: test/tests/build/build_cmake_targets.vader [hunk 7]
```diff
@@ -214,14 +202,8 @@ Execute ([CMake targets] Only default build types have default values):
...
 Execute ([CMake targets] Get default build types):
-    " enable tests for windows with last cmake
-    if !has('win32') || utils#cmake#verNewerOrEq([3, 14])
-        Assert !isdirectory("cmake-build-Release"), "Build directory should not exist"
-        silent CMake
-        Assert filereadable("cmake-build-Release/CMakeCache.txt"), "CMakeCache.txt should be generated"
-        let result = split(cmake4vim#CompleteBuildType(0,0,0), "\n")
-        AssertEqual len(result), 4
-    endif
+    let result = split(cmake4vim#CompleteBuildType(0,0,0), "\n")
```
Comment (ilyachur) [inline test/tests/build/build_cmake_targets.vader:205]: This change shows that something wrong happened. Because before and after each test we remove build folder. It shouldn't work in case if we don't have cached values.

Comment (mark2185) [reply test/tests/build/build_cmake_targets.vader:205]: This returns the default build types even without a build folder, because `cmake` defines them: `[ 'Default', 'Release', 'RelWithDebInfo, 'MinSizeRel' ]`

I removed the `if` because this should work on windows too, and there is no need for running `CMake`.

Comment (ilyachur) [reply test/tests/build/build_cmake_targets.vader:205]: Yes, right. I thought that it is build targets. Sorry.

**Comment by ilyachur** [comment]
@mark2185 Is this PR ready to merge?

**Comment by mark2185** [comment]
> @mark2185 Is this PR ready to merge?

As far as I am concerned - **yes**. I have no idea why did I lose `0.05%` coverage (I saw `codecov`, but I have no idea which test was testing that), but I'm not sure if it's worth enough to investigate.

**Comment by ilyachur** [comment]
> > @mark2185 Is this PR ready to merge?
> 
> As far as I am concerned - **yes**. I have no idea why did I lose `0.05%` coverage (I saw `codecov`, but I have no idea which test was testing that), but I'm not sure if it's worth enough to investigate.

@mark2185 Looks like the red line is not covered. I don't know why your changes affect this line, could you double-check?
Screenshot 2021-08-02 at 15 47 30

**Comment by mark2185** [comment]
> Hmmm, I just think that maybe it can be related with your changes in tests.

I'll try to find out which test it was, then. Probably later today, maybe tomorrow.

**Comment by ilyachur** [comment]
Please rebase your branch to the latest master

File: test/local_run.py [hunk 8]
```diff
@@ -54,10 +54,11 @@ def remove_test_files(root_dir):
             os.environ['VIM_PROFILE_FILE'] = os.path.join(home_dir, 'provile_' + test_case + '_' + args.editor + '_cmake' + cmake_version + '_' + os_name + '.txt')
         start = time.time()
         remove_test_files(current_dir)
-        res = subprocess.run([args.editor, '-Nu', 'vimrc', '+Vader! ' + test_path])
+        res = subprocess.run([args.editor, '-Es', '-Nu', 'vimrc', '+Vader! ' + test_path])
```
Comment (ilyachur) [inline test/local_run.py:57]: Why did you add -Es flags?
